### ISSUE TITLE
Questions regarding the PyTorch tutorial

### ISSUE THREAD

**Issue by netw0rkf10w** [ray-project/ray#10910]
Hello,

Thank you so much for the software!

I've just started learning Ray Tune for PyTorch, and would like to ask some questions related to your [official PyTorch tutorial](https://docs.ray.io/en/master/tune/tutorials/tune-pytorch-cifar.html). 

1. At the beginning of the `train_cifar()` function, we read a checkpoint if it's given:

```python
    if checkpoint_dir:
        checkpoint = os.path.join(checkpoint_dir, "checkpoint")
        model_state, optimizer_state = torch.load(checkpoint)
        net.load_state_dict(model_state)
        optimizer.load_state_dict(optimizer_state)
```

The rest of the tutorial doesn't seem to use this `checkpoint_dir` argument at all, so my first question is: Do these lines of code have some specific purpose?

2. At each epoch in `train_cifar()`, we save a checkpoint:

```python
    for epoch in range(10):  # loop over the dataset multiple times
        # standard PyTorch training and validation phases go here
        ...
        # Ray Tune saves checkpoint here
        with distributed_checkpoint_dir(step=epoch) as checkpoint_dir:
            path = os.path.join(checkpoint_dir, "checkpoint")
            torch.save(
                (net.state_dict(), optimizer.state_dict()), path)

        tune.report(loss=(val_loss / val_steps), accuracy=correct / total)
```

Here `checkpoint_dir` seems to be some directory returned by `distributed_checkpoint_dir` (I'll try to run the code to see which one it is). My second question is: is this `checkpoint_dir` related to the `checkpoint_dir` above (i.e. the optional input argument of `train_cifar()`)?

3. Later in the `main()` function, you do this at the beginning:

```python
    data_dir = os.path.abspath("./data")
    load_data(data_dir)
```
Is there a good reason to `load_data()` here or is it just a mistake?

Thanks in advance for your answers!

**Comment by richardliaw** [comment]
cc @krfricke

**Comment by netw0rkf10w** [comment]
I have been trying very hard to deploy Ray Tune on a Slurm server with PyTorch DDP without success... Apparently this is a very tricky task, as I couldn't find a complete example in the documentation.

In summary, I have a training script which I launch using `torch.distributed.launch`. This script has some input arguments on which I would like to run a hyper-parameter search. Ideally, I would like **each trial to run on a separate node (with 8 GPUs)**, otherwise the search would take too long as my model and dataset are quite large.

The details are below.

To launch a regular training on a single node, I do `sbatch train.slurm` where `train.slurm` looks like the following:

```bash
#!/bin/bash
#SBATCH --ntasks=8
#SBATCH --ntasks-per-node=8
#SBATCH --gres=gpu:8
#SBATCH --cpus-per-task=3
#SBATCH --hint=nomultithread
#SBATCH --distribution=block:block
#SBATCH --time=20:00:00
#SBATCH --exclusive

module purge
module load python/3.7.5 cuda/10.1.2 cudnn/7.6.5.32-cuda-10.1 nccl/2.6.4-1-cuda
 
set -x

OMP_NUM_THREADS=3 python -u -m torch.distributed.launch --nproc_per_node=8 --use_env main.py --num_workers 3 --cfg coco_resnet50.cfg
```

The configuration file `coco_resnet50.cfg` in the above command contains input arguments to the `main.py`. It looks like the following and can be parsed by `argparse` inside `main.py`:

```
backbone=resnet50
dataset=coco
epochs=100
alpha=100.0
beta=0.01
gamma=3.0
...
```

For tuning `alpha, beta, gamma`, I created a new script `main_tune.py` that is almost the same as `main.py`, except the additional Ray code (following [the tutorial](https://docs.ray.io/en/master/tune/tutorials/tune-pytorch-cifar.html)):

```python
import os
import argparse
from pathlib import Path

import torch
from functools import partial
import ray
from ray import tune
from ray.tune import CLIReporter
from ray.tune.schedulers import ASHAScheduler
from ray.tune.integration.torch import (DistributedTrainableCreator,
                                        distributed_checkpoint_dir)

from main import get_args_parser, build_model, build_dataloaders, evaluate, train_one_epoch

def train_function(config, checkpoint_dir=None, args=None):

    args.alpha = config['alpha']
    args.beta = config['beta']
    args.gamma = config['gamma']

    init_distributed_mode(args)
    
    device = torch.device(args.device)
    model = build_model(args)
    model.to(device)

    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu])
    model_without_ddp = model.module

    optimizer = ...

    dataloaders, sampler = build_dataloaders(args)

    if checkpoint_dir:
        resume_path = Path(checkpoint_dir)/'checkpoint.pth'
        if resume_path.is_file():
            checkpoint = torch.load(str(resume_path), map_location='cpu')
            model_without_ddp.load_state_dict((checkpoint['model']))
            optimizer.load_state_dict(checkpoint['optimizer'])

    for epoch in range(args.epochs):
        sampler.set_epoch(epoch)
        train_one_epoch(model, dataloaders, optimizer, device, epoch)
        val_loss, val_acc = evaluate(model, dataloaders, device)

        with distributed_checkpoint_dir(step=epoch) as checkpoint_dir:
            path = os.path.join(checkpoint_dir, "checkpoint.pth")
            torch.save({'model': model_without_ddp.state_dict(),
                                'optimizer': optimizer.state_dict()}, path)

        tune.report(loss= val_loss, accuracy= val_acc)
    

def main(args):

    config = {
        "alpha": tune.loguniform(1, 1e3),
        "beta": tune.loguniform(0.001, 1.0),
        "gamma": tune.loguniform(1, 1e3)
    }
    scheduler = ASHAScheduler(
        metric="accuracy",
        mode="max",
        max_t=args.epochs,
        grace_period=1,
        reduction_factor=2)
    reporter = CLIReporter(
        metric_columns=["loss", "accuracy", "training_iteration"])
    
    # ray.init(address='auto')
    ray.init(address=os.environ["ip_head"])
    print("Nodes in the Ray cluster:")
    print(ray.nodes())

    distributed_train = DistributedTrainableCreator(
        partial(train_function, args=args),
        use_gpu=True,
        num_workers=8,  # number of parallel workers to use
        num_cpus_per_worker=3)
    
    result = tune.run(
        distributed_train,
        resources_per_trial=None,
        config=config,
        num_samples=1,
        scheduler=scheduler,
        progress_reporter=reporter)

    best_trial = result.get_best_trial("accuracy", "max", "last")
    print("Best trial config: {}".format(best_trial.config))
    print("Best trial final validation loss: {}".format(
        best_trial.last_result["loss"]))
    print("Best trial final validation accuracy: {}".format(
        best_trial.last_result["accuracy"]))


if __name__ == '__main__':
    parser = argparse.ArgumentParser('Ray Tune hyperparameter search', parents=[get_args_parser()])
    args = parser.parse_args()
    main(args)
```

Next, I followed the [Deploying on Slurm tutorial](https://docs.ray.io/en/master/cluster/slurm.html), or more precisely @pengzhenghao's [template script](https://gist.github.com/pengzhenghao/b348db1075101a9b986c4cdfea13dcd6) (I also tried the provided Starter SLURM script but it didn't work), and created a new `tune.slurm` as follows:

```bash
#!/bin/bash
#SBATCH --ntasks=8
#SBATCH --ntasks-per-node=8
#SBATCH --gres=gpu:8
#SBATCH --cpus-per-task=3
#SBATCH --hint=nomultithread
#SBATCH --distribution=block:block
#SBATCH --time=20:00:00
#SBATCH --exclusive

module purge
module load python/3.7.5 cuda/10.1.2 cudnn/7.6.5.32-cuda-10.1 nccl/2.6.4-1-cuda
 
set -x

# ===== Define variables =====
worker_num=1
num_gpus=8  # per node
num_cpus=3 # per node

nodes=$(scontrol show hostnames $SLURM_JOB_NODELIST) # Getting the node names
nodes_array=($nodes)

echo "Current nodes:"
echo $nodes

node1=${nodes_array[0]}

echo "Head Node:"
echo $node1

ips=$(srun --nodes=1 --ntasks=1 -w $node1 hostname --ip-address) # Making address

echo "ips: "
echo $ips

ip_array=($ips)
ip_prefix=${ip_array[0]} # The first element is ipv6 address

suffix=':6379'
ip_head=$ip_prefix$suffix
redis_password=$(uuidgen)

export ip_head # Exporting for latter access by your command
export redis_password # Exporting for latter access by your command
echo "ip_head"
echo $ip_head

# ===== Start the head node =====
srun --nodes=1 --ntasks=1 -w $node1 ray start --block --head --redis-port=6379 --redis-password=$redis_password --num-gpus=$num_gpus --num-cpus=$num_cpus &# Starting the head
sleep 30

# ===== Start worker node =====
for ((i = 1; i <= $worker_num; i++)); do
  node2=${nodes_array[$i]}
  srun --nodes=1 --ntasks=1 -w $node2 ray start --block --address=$ip_head --redis-password=$redis_password --num-gpus=$num_gpus --num-cpus=$num_cpus &# Starting the workers
  sleep 30
done

# ===== Submit task =====

OMP_NUM_THREADS=3 python -u main_tune.py --num_workers 3 --cfg coco_resnet50.cfg
```

Submitting this job, I obtained this error:

```
srun: Job 320376 step creation temporarily disabled, retrying
srun: Step created for job 320376
slurmstepd: error: execve(): ray: No such file or directory
srun: error: [node_name]: task 0: Exited with exit code 2
srun: Terminating job step 320376.1
```

Could you please tell me what I did wrong?

Thank you very much in advance for your help!

**Comment by krfricke** [comment]
Hi,

regarding your first post:

1) The `checkpoint_dir` parameter is automatically populated by Tune if the scheduler decides to restore a checkpoint. The code is then used to restore model weights/parameters from this checkpoint. This could happen e.g. if you continue trials after shutting down the training, or if algorithms such as PBT choose to copy parameters from one trial to another.

2) Yes, this is the part of the code where checkpoints are saved. By using `tune.checkpoint_dir` or distributed_checkpoint_dir` the checkpoint gets automatically registered with Tune. These registered checkpoints can then be passed to the train function. As mentioned above, Tune's schedulers usually make these decisions.

3) In the example we are starting several trials, all of which need access to the same data. In the example, we download this data before starting the run to avoid that each trial tries to download the data (which would lead to multiple concurrent downloads and potentially conflicts, as no lock file is used here). For distributed settings with multiple nodes this is less important, especially if you only run one trial per machine. You can then remove this part of the code.

I'm not really familiar with Tune on Slurm. It seems however that the ray library has not been installed on the machine. `ray` is our command line interface script, and it should be automatically available when ray gets installed via `pip`.

**Comment by netw0rkf10w** [comment]
Hi @krfricke,

Thank you very much for your answers!

> 1. The `checkpoint_dir` parameter is automatically populated by Tune if the scheduler decides to restore a checkpoint. The code is then used to restore model weights/parameters from this checkpoint. This could happen e.g. if you continue trials after shutting down the training, or if algorithms such as PBT choose to copy parameters from one trial to another.

I see. I would suggest to add the above explanation to the tutorial (e.g. right after the sentence "The checkpoint_dir parameter is used to restore checkpoints."), for the sake of inexperienced users like me.

> 3. In the example we are starting several trials, all of which need access to the same data. In the example, we download this data before starting the run to avoid that each trial tries to download the data (which would lead to multiple concurrent downloads and potentially conflicts, as no lock file is used here). For distributed settings with multiple nodes this is less important, especially if you only run one trial per machine. You can then remove this part of the code.

That totally makes sense. I once used this trick to pre-download data, but for some reason I didn't recognize it when reading the tutorial.
 
> I'm not really familiar with Tune on Slurm. It seems however that the ray library has not been installed on the machine. `ray` is our command line interface script, and it should be automatically available when ray gets installed via `pip`.

I have `ray` installed via `pip` though... Could you please tell me whom I should ask Slurm related questions?

Thank you again!

**Comment by richardliaw** [comment]
@netw0rkf10w can you open a separate issue for SLURM related questions? I can forward that to those who are more familiar. 

```
slurmstepd: error: execve(): ray: No such file or directory
```
this may not be seeing pip; try running `which -a ray` too.

Also, try running:
```
which pip
which python
python -c 'import ray; print(ray.__file__)'
```

I will close this issue but be sure to open a new one ^ with the above information!

**Comment by netw0rkf10w** [comment]
@richardliaw Thanks. Finally I could make it work, but still I'm not sure if the behavior is correct. I've just posted a new [issue](https://github.com/ray-project/ray/issues/10986), please have look.

**Comment by netw0rkf10w** [comment]
@krfricke Could you please tell me if there's an option to save the current Tune job state and resume it later? My Slurm server has a time limit for each job, and I'm afraid that tuning large models will surpass that time limit. Thanks a lot!

**Comment by krfricke** [comment]
Hi @netw0rkf10w, absolutely.

There's three things you'll need to make this work:

1. Enable checkpointing (otherwise you cannot continue the training)
2. Call `tune.run()` with the `resume` parameter
3. Provide a logdir (this is actually optional but it makes it easier to track your run status)

Here's a simple example:

```
import os
import time

from ray import tune
import pickle


def train(config, checkpoint_dir=None):
    if checkpoint_dir:
        with open(os.path.join(checkpoint_dir, "val.ckpt"), "rb") as fp:
            step, random = pickle.load(fp)
            print(f"LOADED CHECKPOINT WITH {step} AND {random}")
    else:
        step = 0
        random = config["random"]

    base = config["base"]

    for i in range(step, 20):
        score = base + step
        if i % 2 == 0 or True:
            with tune.checkpoint_dir(i) as checkpoint_dir:
                with open(os.path.join(checkpoint_dir, "val.ckpt"), "wb") as fp:
                    pickle.dump((i, random), fp)
        tune.report(score=score)
        time.sleep(1)


if __name__ == "__main__":
    # This is only needed for this simple example because the experiment
    # state is only saved every 10 seconds otherwise. For long running
    # experiments, this should be unnecessary
    os.environ["TUNE_GLOBAL_CHECKPOINT_S"] = "1"
    
    config = {
        "base": tune.grid_search([100, 200, 300, 400, 500]),
        "random": tune.randint(1000, 2000)
    }

    logdir = "/tmp/release_continue"

    if not os.path.isdir(logdir):
        os.mkdir(logdir, 0o755)

    tune.run(
        train,
        config=config,
        local_dir=logdir,
        num_samples=4,
        resume="PROMPT"
    )

```

Can you work out on how to adapt this to your use case yourself?

**Comment by netw0rkf10w** [comment]
@krfricke Awesome!

According to the documentation:

> resume (str|bool) – One of “LOCAL”, “REMOTE”, “PROMPT”, “ERRORED_ONLY”, or bool. LOCAL/True restores the checkpoint from the local_checkpoint_dir, determined by name and local_dir...

if I set `resume='LOCAL'` or `resume=True` then it will resume from `local_checkpoint_dir`, "determined by `name` and `local_dir`". What happens if I don't set a `name` to the experiment?

My Python code looks like this:

```python
    local_dir = '~/ray_results'
    if args.output_dir:
        os.makedirs(args.output_dir, exist_ok=True)
        local_dir = args.output_dir

    print(f'Will save Tune results to {local_dir}')

    result = tune.run(
        distributed_train,
        resources_per_trial=None,
        config=config,
        num_samples=args.num_trials,
        scheduler=scheduler,
        progress_reporter=reporter,
        local_dir=local_dir)
```

I didn't think about setting a name. Will it be able to resume if I set `resume=True` though? 

Thanks again!

**Comment by krfricke** [comment]
Yes, you will. If you don't specify a name, the default name will be determined by the name of the trainable function. E.g. in your case it will be `distributed_train`.
